feat(bridge): use real bridge provider

shoom3301 opened the pull request:
# Summary

This PR is very coupled with the SDK changes: https://github.com/cowprotocol/cow-sdk/pull/275

## Changes

1. Replaced mocked `testBridgeProvider` with `acrossBridgeProvider` from SDK
2. Replaced local `ChainInfo` state with one from SDK
3. Added trade params resetting when you had a buy order setup and pick a bridge buy token
4. Added a custom signer (static private key) for bridge quote fetching, because it requires hooks signing
5. Added very dummy `getErc20Tokens` function which must be changed in the future: https://www.notion.so/cownation/Fetch-token-details-1dc8da5f04ca80e6a468c5da9455184a?pvs=4
6. Adjusted quote polling mechanism in order to support bridge quotes
7. Adjusted BridgeSDK in order to support orders with pre hooks (permit)

# To Test

**This is not a final version of Bridging feature!**

1. Non bridge orders should work as before
2. Bridge trades should work (getting quote, posting an order with corresponding post-hook)

Please, pay attention to different order parameters such:
 - validTo
 - partiallyFillable
 - receiver
 - appData
 - kind

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced cross-chain trading support using a new bridging SDK, enabling seamless bridging and swapping across multiple networks.
  - Added utility for fetching ERC-20 token information and decimals across supported chains.
  - Enhanced trade quote management with dedicated handling for bridge quotes and related parameters.
  - Added a static signer utility for bridge quote signing on supported chains.

- **Improvements**
  - Updated network and token data structures for richer chain and token metadata, improving accuracy and display consistency.
  - Improved navigation and trade state logic to better handle bridging scenarios and prevent unsupported actions.
  - Expanded token list and chain selector components with more detailed network information and logos.
  - Refined trade widget input disabling logic to include bridging state.
  - Enhanced quote parameter construction and error handling in quote fetching.
  - Updated token and chain imports for consistency and simplification.
  - Adjusted chain selection logic to better handle bridge-supported networks.
  - Replaced mock bridge provider with production SDK-based provider for live data.
  - Improved token mapping and error handling in bridge-supported tokens retrieval.
  - Exported additional types for improved type safety in app data handling.

- **Bug Fixes**
  - Enhanced error handling and logging for quote fetching and unsupported token scenarios.
  - Improved type safety and consistency in trade and quote parameter handling.
  - Fixed token logo and property mapping issues in bridge-supported tokens.

- **Chores**
  - Upgraded dependencies to the latest SDK version for improved compatibility and features.
  - Cleaned up and refactored import statements and removed unused mock and type files.
  - Consolidated and simplified import statements across modules.
  - Updated test mocks to reflect SDK changes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeQuote/services/fetchAndProcessQuote.ts:
```diff
@@ -43,7 +47,13 @@ export async function fetchAndProcessQuote(
-    tradeQuoteManager.onResponse(data, fetchParams)
+    tradeQuoteManager.onResponse(
+      {
+        quoteResults: (data as any).swap || (data as any).quoteResults, // TODO fix types
```
Need to fix types and save `BridgeQuoteResults` into jotai atom as well, see `BridgeQuoteAndPost['bridge']` in SDK

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeQuote/utils/quoteUsingSameParameters.ts:
```diff
@@ -14,28 +14,35 @@ import { AppDataInfo } from 'modules/appData'
+   * TODO: in bridging mode, appData contains only hooks without other information
+   * TODO: in bridging mode, receiver became a CoW Shed address
+   */
+  const isBridge = true
```
This is a bit of hell...
This function is responsible for skiping quote fetching when nothing real is changed.
With briding quotes it's a bit different than with CoW Protocol quotes.
1. AppData comparation doesn't work. Probably it's related to https://github.com/cowprotocol/cow-sdk/pull/275/files#r2024716345. Need to check why it's different
2. Bridge quote changes buyToken to intermediate one. As I mentioned [here](https://github.com/cowprotocol/cowswap/pull/5588/files#r2024702243), we need to save `BridgeQuoteAndPost['bridge']` into `tradeQuotesAtom` and use the state here in order to properly compare buy tokens.
3. Bridge quote changes to CoW Shed, we also need to handle that fact here

alfetopito commented on apps/cowswap-frontend/src/modules/trade/hooks/useNavigateOnCurrencySelection.ts:
```diff
@@ -80,12 +81,22 @@ export function useNavigateOnCurrencySelection(): CurrencySelectionCallback {
        */
       const shouldResetBuyToken = isInputField && targetChainMismatch
       const shouldSetTargetChain = !isInputField && targetChainMismatch
+      const shouldResetBuyOrder = !isInputField && orderKind === OrderKind.BUY
```
Shouldn't this be done only if this is a bridge swap?

shoom3301 replied:
Nice catch! Thanks!
https://github.com/cowprotocol/cowswap/pull/5588/commits/e05acf33b066a303afa3d23159a3b6b937fe8e0c

alfetopito commented on apps/cowswap-frontend/src/modules/tradeQuote/utils/quoteUsingSameParameters.ts:
```diff
@@ -14,28 +14,35 @@ import { AppDataInfo } from 'modules/appData'
+  // TODO: compare tokens chainIds
   return (
-    hasSameAppData &&
+    (isBridge ? true : hasSameAppData) &&
```
Nitpick:

```suggestion
    (isBridge || hasSameAppData) &&
```

Same below

shoom3301 replied:
Completely changed that part

shoom3301 commented on apps/cowswap-frontend/src/tradingSdk/getErc20Tokens.ts:
```diff
@@ -0,0 +1,41 @@
+
+  tokensCache.set(chainId, tokensByAddress)
+
+  return getTokensFromLists(addresses, tokensByAddress)
```
This is very limited mechanism, definitely must be changed.
https://cowservices.slack.com/archives/C0361CDG8GP/p1745239081331449

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeQuote/hooks/useQuoteParams.ts:
```diff
@@ -56,21 +60,32 @@ export function useQuoteParams(amount: Nullish<string>): QuoteParams | undefined
+    const quoteParams: QuoteBridgeRequest = {
       kind: orderKind,
-      sellToken,
-      buyToken,
+      amount: BigInt(amount),
+      owner,
+
+      sellTokenChainId: inputCurrency.chainId,
+      sellTokenAddress,
       sellTokenDecimals,
+
+      buyTokenChainId: outputCurrency.chainId,
+      buyTokenAddress,
       buyTokenDecimals,
-      owner: (account || ZERO_ADDRESS) as `0x${string}`,
-      amount,
+
+      account: owner,
+      appCode,
+      signer: provider.provider || provider.getSigner(),
+
       receiver,
       validFor: DEFAULT_QUOTE_TTL,
       ...(volumeFee ? { partnerFee: volumeFee } : null),
     }
```
The same values, but different goals

alfetopito reviewed (commented):
Reviewed the code, testing now

alfetopito commented:
1. Auto slippage is not respected for bridging (worked fine for regular)

At least it's not reflected in the appData. [Order](https://dev.explorer.cow.fi/orders/0xce9a2c91cb1c28c2df18fc7950fd7a29bdfd12afdea8b1749942b070fb0966d05b0abe214ab7875562adee331deff0fe1912fe4268091c97).

Auto-slippage suggested was 5%.
Appdata shows 50 bps => 0.5%

Placed [another order](https://dev.explorer.cow.fi/orders/0xadba5717ea786b3587ff77596f13e763d514bf5eec048b9e9250d44cfbc30c715b0abe214ab7875562adee331deff0fe1912fe4268091e5a) and confirmed. Limit order slippage that's signed remained at 0.5%.

![image](https://github.com/user-attachments/assets/dd77e124-c2c3-4478-93bf-6d0b11beaa86)

shoom3301 commented on apps/cowswap-frontend/src/modules/tradeQuote/utils/quoteUsingSameParameters.ts:
```diff
@@ -57,7 +79,34 @@ function compareAppDataWithoutQuoteData(a: AppDataInfo['doc'] | undefined, b: Ap
 function removeQuoteMetadata(appData: AppDataInfo['doc']): string {
-  const { metadata: fullMetadata, ...rest } = appData
-  const { quote: _, ...metadata } = fullMetadata
+  const { metadata: fullMetadata, environment: _e, ...rest } = appData
```
Changed this: https://github.com/cowprotocol/cowswap/pull/5588/commits/6733d8e1135fafc4c027d6ca560b720c2b1bcbbc

shoom3301 commented:
> 1. Auto slippage is not respected for bridging (worked fine for regular)
> 
> At least it's not reflected in the appData. [Order](https://dev.explorer.cow.fi/orders/0xce9a2c91cb1c28c2df18fc7950fd7a29bdfd12afdea8b1749942b070fb0966d05b0abe214ab7875562adee331deff0fe1912fe4268091c97).
> 
> Auto-slippage suggested was 5%. Appdata shows 50 bps => 0.5%
> 
> Placed [another order](https://dev.explorer.cow.fi/orders/0xadba5717ea786b3587ff77596f13e763d514bf5eec048b9e9250d44cfbc30c715b0abe214ab7875562adee331deff0fe1912fe4268091e5a) and confirmed. Limit order slippage that's signed remained at 0.5%.

Nice catch @alfetopito !
Fixed: https://github.com/cowprotocol/cow-sdk/pull/294

shoom3301 commented:
@elena-zh thank you!

1. Out of scope. Should be done as part of https://www.notion.so/cownation/Fetch-token-details-1dc8da5f04ca80e6a468c5da9455184a?pvs=4
2. Fixed
3. I cannot reproduce it now. Might be a temporary problem?
4. Fixed. Added an ugly error, it should be perttified in the next iterations
5. Out of scope. Need to address separately.
6. Out of scope. Need to address ETH-flow separately.
7. Out of scope. Need to address ETH-flow separately.
8. Out of scope. Need to address ETH-flow separately.
9. Out of scope
10. Out of scope
11. Out of scope
12. Out of scope
13. Out of scope
14. Out of scope

Sorry for so many "Out of scope", but I can't do everything in one PR.
Scope of this PR is only BridgingSDK intergration (get quote, sign and send order with post-hook.)

elena-zh commented:
> @elena-zh thank you!
> 
> 1. Out of scope. Should be done as part of https://www.notion.so/cownation/Fetch-token-details-1dc8da5f04ca80e6a468c5da9455184a?pvs=4
> 2. Fixed
> 3. I cannot reproduce it now. Might be a temporary problem?
> 4. Addressed in [feat(bridge): display quote error and details #5633](https://github.com/cowprotocol/cowswap/pull/5633). Added an ugly error, it should be perttified in the next iterations
> 5. Out of scope. Need to address separately.
> 6. Out of scope. Need to address ETH-flow separately.
> 7. Out of scope. Need to address ETH-flow separately.
> 8. Out of scope. Need to address ETH-flow separately.
> 9. Out of scope
> 10. Out of scope
> 11. Out of scope
> 12. Out of scope
> 13. Out of scope
> 14. Out of scope
> 
> Sorry for so many "Out of scope", but I can't do everything in one PR. Scope of this PR is only BridgingSDK intergration (get quote, sign and send order with post-hook.)

Might it be possible to write a better description then?

e.g. `smoke test: place an order with ERC-20 tokens and verify that it is executed.`

alfetopito commented on apps/cowswap-frontend/src/modules/tradeQuote/utils/quoteUsingSameParameters.ts:
```diff
@@ -50,6 +72,20 @@ function compareAppDataWithoutQuoteData(a: AppDataInfo['doc'] | undefined, b: Ap
     return a === b
   }
 
+  // When environment is set in one appData and not in the other, we need to remove it
```
What I don't get is, why/when is it not set?

shoom3301 replied:
🤦 
Sorry, I found and fixed the cause:
https://github.com/cowprotocol/cow-sdk/pull/294/commits/9024a44c18f5c7919d6e948282391e975b3ac54e

Thanks for insisting on it!

elena-zh commented:
Hey @shoom3301 , the build fails. Could you please check what is wrong there?

shoom3301 commented:
@elena-zh sorry, should be fixed now
